### PR TITLE
refactor(control_validator): reduce copy and add test for trajectory aligning

### PR DESCRIPTION
**Pull request by paulsohn**
## Description
Refactor `align_trajectory_with_reference_trajectory` to reduce copy, and introduce (copilot-generated) tests.

This aims to solve the same control validator crash issue with #10372 and #11682 (my previous PR), but unfortunately it is reported that #11682 does not fix this issue - suggesting that we need more aggresive refactoring.

This work is done on top of #11682 , which includes:
> * The arguments and return type of `align_trajectory_with_reference_trajectory` are no longer `Trajectory` message type but `std::vector<TrajectoryPoint>`, which prevents copying on type conversion.
> * Reserve-and-push-back patterns are simplified to iterator-based construction.

Besides renaming for more clarity, a few points are improved:
* manipulate aligned trajectory point vectors in its reversed form, so that we can avoid costly pushing and popping at the front.
* for trajectory aligning (clamping) purpose, we do not need the entire reference trajectory in this algorithm.
  * the only relevant segments are the first and the last ones in the reference trajectory, so maintain only those instead of copying or clamping the whole.
  * likewise, when we calculate the interpolated point on the predicted trajectory, we already know which segment to interpolate. `remove_front_trajectory_point` and `insert_point_in_predicted_trajectory` are merged to pipeline that exact segment.

The only $O(N)$ copy / memory allocations involved in this function are now:
* limiting the predicted trajectory size to `MAX_TRAJECTORY_SIZE` (hence $N \leq 5000$; this might be replaced by iterator pairs or spans)
* reversing twice the in-progress aligned trajectory

Obviously, we can further refactor this into single $O(N)$ copy (or even zero!), no reversal, if we use two iterators from the beginning and the end. But I am afraid that it would lead to bug-prone and less maintainable code, so I didn't do here.

## Related links

* #10372 #11682 
* TIER IV internal: [T4DEV-34730](https://tier4.atlassian.net/browse/T4DEV-34730) [T4DEV-40047](https://tier4.atlassian.net/browse/T4DEV-40047)

## How was this PR tested?
`align_trajectory_with_reference_trajectory` is now unit-test included. Simple cases are covered, and one pathological one which is disabled by default (see notes below)

```bash
$ colcon build --symlink-install --packages-select autoware_control_validator
$ colcon test --packages-select autoware_control_validator
```

If the pathological test case is enabled, then the test fails.

- [x] Passed TIERIV's internal CI checks
  - [x] [Test link part 1](https://evaluation.tier4.jp/evaluation/reports/b5480e5d-da86-5afd-a8b5-c97b00aee0cd?project_id=prd_jt)
  - [x] [Test link part 2](https://evaluation.tier4.jp/evaluation/reports/11432ac1-9032-594e-8103-4ba7821679ad?project_id=prd_jt)
  - [x] [Test link part 3](https://evaluation.tier4.jp/evaluation/reports/0026edd7-ee06-517c-b79c-705cd3acf969?project_id=prd_jt)

## Notes for reviewers

While reviewing the original algorithm, I found a 'pathological' edge case:

image

If the reference and predicted trajectories are given as the above figure, the aligned trajectory is empty since neither the first nor the last segment of the reference trajectory overlap with the aligned trajetory, despite the middle segment does overlap.

I am no planning/control expert, so I cannot tell how realistic edge cases can be implied from this exaggerated example. If this phenomenon may cause any real-world issue, then it would worth proposing a better algorithm that embraces this kind of cases. I added the 'expected' behavior as a disabled test case, so that it can be enabled on core algorithm revision.

That said, this note is unrelated to the PR itself, since the purpose of the PR is just to provide an optimization of the existing algorithm so that the node crashing problem can be resolved quickly without breaking any functional behaviors.

## Interface changes

None.

## Effects on system behavior

Less copies and memory safety guards. Each direct dereferencing memory accesses (`.back()`, `.front()`, `.pop_back()`, indexing on vectors, `.value()` on optional) at the point of 1c0e4b2b27dd3e38a7ddbf8a761fbd346e8b266f has a explicit guard before it.

### CODE REVIEW

File: control/autoware_control_validator/src/utils.cpp [hunk 1]
```diff
@@ -19,176 +19,171 @@
...
+  clamp_trajectory_head(
+    first_segment_points, reversed_aligned_trajectory_points, predicted_trajectory_points);
+
+  if (reversed_aligned_trajectory_points.empty()) {
+    return TrajectoryPoints();
   }
 
-  // If last point of predicted_trajectory is behind of end of trajectory, erase points which are
-  // behind trajectory last point and insert pNew along the predicted_trajectory
+  // If last point of predicted_trajectory is behind of reference_trajectory, remove points which
```
[the remark this answers is not in the document]
Comment (paulsohn) [reply control/autoware_control_validator/src/utils.cpp]: a35746dd610fd90682ab775b388fa44902855478

File: control/autoware_control_validator/test/test_utils.cpp [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  auto quat = tf2::toMsg(q);
+  for (size_t i = 0; i < large_n; ++i) {
+    TrajectoryPoint p;
+    p.pose.position.x = static_cast<float>(i * 0.01);
+    p.pose.position.y = 0.0f;
+    p.pose.orientation = quat;
+    pred.emplace_back(p);
+  }
+  // reference adequate
+  auto ref = make_linear_trajectory_range(0.0, 100.0, 101.0);
```
[the remark this answers is not in the document]
Comment (paulsohn) [reply control/autoware_control_validator/test/test_utils.cpp]: a35746dd610fd90682ab775b388fa44902855478

File: control/autoware_control_validator/test/test_utils.cpp [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    EXPECT_NEAR(out[i].pose.position.x, pred[i].pose.position.x, 1e-6);
+    EXPECT_NEAR(out[i].pose.position.y, pred[i].pose.position.y, 1e-6);
+  }
+}
+
+TEST(align_trajectory_with_reference_trajectory, NonOverlappingBeforeAfter_ReturnsEmpty)
+{
+  // predicted entirely before reference
+  auto ref = make_linear_trajectory_range(0.0, 10.0, 11.0);
+  auto pred_before = make_linear_trajectory_range(-20.0, -10.0, 11.0);
```
[the remark this answers is not in the document]
Comment (paulsohn) [reply control/autoware_control_validator/test/test_utils.cpp]: a35746dd610fd90682ab775b388fa44902855478

File: control/autoware_control_validator/test/test_utils.cpp [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+  auto out_before = align_trajectory_with_reference_trajectory(ref, pred_before);
+  EXPECT_TRUE(out_before.empty());
+
+  // predicted entirely after reference
+  auto pred_after = make_linear_trajectory_range(20.0, 30.0, 11.0);
```
[the remark this answers is not in the document]
Comment (paulsohn) [reply control/autoware_control_validator/test/test_utils.cpp]: a35746dd610fd90682ab775b388fa44902855478

File: control/autoware_control_validator/test/test_utils.cpp [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+  auto ref2 = make_linear_trajectory_range(0.0, 10.0, 11.0);
+  auto pred2 = make_linear_trajectory({});
+  auto out2 = align_trajectory_with_reference_trajectory(ref2, pred2);
+  EXPECT_TRUE(out2.empty());
+}
+
+TEST(align_trajectory_with_reference_trajectory, IdenticalTrajectories_ReturnsSame)
+{
+  auto ref = make_linear_trajectory_range(0.0, 10.0, 11.0);
+  auto pred = make_linear_trajectory_range(0.0, 10.0, 11.0);
```
[the remark this answers is not in the document]
Comment (paulsohn) [reply control/autoware_control_validator/test/test_utils.cpp]: a35746dd610fd90682ab775b388fa44902855478

File: control/autoware_control_validator/test/test_utils.cpp [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+  // predicted empty -> empty
+  auto ref2 = make_linear_trajectory_range(0.0, 10.0, 11.0);
+  auto pred2 = make_linear_trajectory({});
+  auto out2 = align_trajectory_with_reference_trajectory(ref2, pred2);
+  EXPECT_TRUE(out2.empty());
+}
+
+TEST(align_trajectory_with_reference_trajectory, IdenticalTrajectories_ReturnsSame)
+{
+  auto ref = make_linear_trajectory_range(0.0, 10.0, 11.0);
```
[the remark this answers is not in the document]
Comment (paulsohn) [reply control/autoware_control_validator/test/test_utils.cpp]: a35746dd610fd90682ab775b388fa44902855478

File: control/autoware_control_validator/test/test_utils.cpp [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+  auto out = align_trajectory_with_reference_trajectory(ref, pred);
+  EXPECT_LE(out.size(), 5000u);
+}
+
+TEST(
+  align_trajectory_with_reference_trajectory,
+  PartialOverlap_StartInsertionProducesInterpolatedPoint)
+{
+  // reference from 0..10
+  auto ref = make_linear_trajectory_range(0.0, 10.0, 11.0);
```
[the remark this answers is not in the document]
Comment (paulsohn) [reply control/autoware_control_validator/test/test_utils.cpp]: a35746dd610fd90682ab775b388fa44902855478

File: control/autoware_control_validator/test/test_utils.cpp [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+TEST(align_trajectory_with_reference_trajectory, EmptyReferenceOrPredicted)
+{
+  // reference with fewer than 2 points -> empty
+  auto ref1 = make_linear_trajectory({0.0});
+  auto pred1 = make_linear_trajectory_range(0.0, 10.0, 11.0);
+  auto out1 = align_trajectory_with_reference_trajectory(ref1, pred1);
+  EXPECT_TRUE(out1.empty());
+
+  // predicted empty -> empty
+  auto ref2 = make_linear_trajectory_range(0.0, 10.0, 11.0);
```
[the remark this answers is not in the document]
Comment (paulsohn) [reply control/autoware_control_validator/test/test_utils.cpp]: a35746dd610fd90682ab775b388fa44902855478

File: control/autoware_control_validator/test/test_utils.cpp [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+  for (size_t i = 0; i < out.size(); ++i) {
+    EXPECT_NEAR(out[i].pose.position.x, pred[i].pose.position.x, 1e-6);
+    EXPECT_NEAR(out[i].pose.position.y, pred[i].pose.position.y, 1e-6);
+  }
+}
+
+TEST(align_trajectory_with_reference_trajectory, NonOverlappingBeforeAfter_ReturnsEmpty)
+{
+  // predicted entirely before reference
+  auto ref = make_linear_trajectory_range(0.0, 10.0, 11.0);
```
[the remark this answers is not in the document]
Comment (paulsohn) [reply control/autoware_control_validator/test/test_utils.cpp]: a35746dd610fd90682ab775b388fa44902855478

File: control/autoware_control_validator/test/test_utils.cpp [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+    points.emplace_back(p);
+  }
+  return points;
+}
+
+TEST(align_trajectory_with_reference_trajectory, EmptyReferenceOrPredicted)
+{
+  // reference with fewer than 2 points -> empty
+  auto ref1 = make_linear_trajectory({0.0});
+  auto pred1 = make_linear_trajectory_range(0.0, 10.0, 11.0);
```
[the remark this answers is not in the document]
Comment (paulsohn) [reply control/autoware_control_validator/test/test_utils.cpp]: a35746dd610fd90682ab775b388fa44902855478

**Comment by kyoichi-sugahara** [comment]
@paulsohn 
~~In addition to the bug fixes, the usage of `detail` is now consistent, and points that should have been organized previously have been corrected, which is excellent. Thank you very much.
The test creation is also very helpful.
Would it be possible to position the newly added test file as a test for TrajectoryValidator, create a file named test_trajectory_validator?, and move the TrajectoryValidator related tests from test_control_validator.cpp to the new file?~~
Please ignore about above commnet :bow: 
this test is for util function and we can keep as it is

**Comment by kyoichi-sugahara** [comment]
@paulsohn 
> I added this as a test case to describe where the current algorithm might not work as expected, not because I am asserting this is the intended behavior and must pass even after the core algorithm revision. One may drop this specific test case or update the expected output in the future.

For the edge case you provided, I believe the ideal expected behavior is to return the overlapping segment.
Since this is likely a rare scenario, it doesn't require an immediate fix. However, could you please define the test expectation to return the overlapping path, and then mark the test as DISABLED?

**Comment by paulsohn** [comment]
> However, could you please define the test expectation to return the overlapping path, and then mark the test as DISABLED?

@kyoichi-sugahara Done 212f5fed7bfcc009b3dacfc06fe1cd434af69ea5

File: control/autoware_control_validator/src/utils.cpp [hunk 11]
```diff
@@ -19,176 +19,158 @@
...
-
-  return predicted_trajectory_point_removed;
+  if (reversed_trajectory_points.empty() || !opt_last_removed_point.has_value()) {
+    // boundary segment does not exist and nothing to insert; either the predicted trajectory is
+    // fully consumed, or the referenced segment is ahead of the predicted trajectory
+    return;
+  }
+  // we don't need the whole segments of the predicted trajectory for interpolation
+  TrajectoryPoints boundary_segment_points = {
+    opt_last_removed_point.value(), reversed_trajectory_points.back()};
```
[the remark this answers is not in the document]
Comment (paulsohn) [reply control/autoware_control_validator/src/utils.cpp:72]: Because `reversed_trajectory_points` is given in the reversed order of the intended direction (as the name suggests), this is not true.

File: control/autoware_control_validator/test/test_utils.cpp [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+static TrajectoryPoints make_trajectory_range(
+  std::pair<double, double> start, std::pair<double, double> end, size_t n)
+{
+  TrajectoryPoints points;
+  points.reserve(n);
+  tf2::Quaternion q;
+  q.setRPY(0, 0, 0);
+  auto quat = tf2::toMsg(q);
```
[the remark this answers is not in the document]
Comment (paulsohn) [reply control/autoware_control_validator/test/test_utils.cpp:60]: won't do nitpick

File: control/autoware_control_validator/test/test_utils.cpp [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+
+#include <gtest/gtest.h>
+
+#include <cmath>
+#include <utility>
+#include <vector>
+
+using autoware_planning_msgs::msg::Trajectory;
+using autoware_planning_msgs::msg::TrajectoryPoint;
+
```
[the remark this answers is not in the document]
Comment (paulsohn) [reply control/autoware_control_validator/test/test_utils.cpp:30]: seems plausible, but won't do nitpick

File: control/autoware_control_validator/src/utils.cpp [hunk 14]
```diff
@@ -19,176 +19,158 @@
...
+namespace detail
 {
-  const auto point_to_interpolate = autoware::motion_utils::calcInterpolatedPoint(
-    convertToTrajectory(predicted_trajectory), reference_pose);
-  modified_trajectory.insert(modified_trajectory.begin(), point_to_interpolate);
-}
 
-TrajectoryPoints reverse_trajectory_points(const TrajectoryPoints & trajectory_points)
+inline TrajectoryPoints reverse_trajectory_points(const TrajectoryPoints & trajectory_points)
 {
```
[the remark this answers is not in the document]
Comment (paulsohn) [reply control/autoware_control_validator/src/utils.cpp:35]: won't do nitpick

File: control/autoware_control_validator/src/utils.cpp [hunk 15]
```diff
@@ -19,176 +19,158 @@
...
+/**
+ * @brief Remove trajectory points which are in front of the reference segment
+ * and insert interpolated point at the boundary if needed.
+ *
+ * To avoid front removal and insertion, trajectory is given in reversed order.
+ * @param[in] segment_points reference segment trajectory of length at least 2
+ * @param[inout] reversed_trajectory_points trajectory in reversed order, possibly empty
+ */
+void clamp_trajectory_head(
+  const TrajectoryPoints & segment_points, TrajectoryPoints & reversed_trajectory_points)
```
[the remark this answers is not in the document]
Comment (paulsohn) [reply control/autoware_control_validator/src/utils.cpp:48]: won't do nitpick

File: control/autoware_control_validator/test/test_utils.cpp [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+  //
+  // even though the middle reference segment aligns with predicted trajectory,
+  // current algorithm only considers the first and last segments for overlap checking,
+  // so the result is empty.
+
+  ASSERT_EQ(out.size(), pred.size());
+  for (size_t i = 0; i < out.size(); ++i) {
+    EXPECT_NEAR(out[i].pose.position.x, pred[i].pose.position.x, 1e-6);
+    EXPECT_NEAR(out[i].pose.position.y, pred[i].pose.position.y, 1e-6);
+  }
```
[the remark this answers is not in the document]
Comment (paulsohn) [reply control/autoware_control_validator/test/test_utils.cpp:205]: True, but this is intended as long as the test case is disabled.
